### DISCUSSION TITLE
Lesson 4: npm install --global yarn

### DISCUSSION THREAD

**Discussion by CrispyButter** [smartcontractkit/full-blockchain-solidity-course-py#469] [Q&A] [answered]
Hey everyone I know this may be a fairly simple fix but I am having issues when trying to install yarn. I downloaded nodejs version 16.13.1 LTS and verified it in the terminal but when I run npm install --global yarn I receive the below. Any help is appreciated so I can continue the course.

(venv) Trell@Kentrells-MacBook-Pro web3_py_simple_storage % npm install --global yarn
npm ERR! code EACCES
npm ERR! syscall mkdir
npm ERR! path /usr/local/lib/node_modules/yarn
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/yarn'
npm ERR!  [Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/yarn'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'mkdir',
npm ERR!   path: '/usr/local/lib/node_modules/yarn'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/Trell/.npm/_logs/2021-12-01T23_17_25_582Z-debug.log

**Comment by cromewar** [comment]
hello @CrispyButter it says you don't have permissions to install npm packages globally so there are 3 alternatives for solution.

1. install yarn just for each project, yo must remove the `--global` modifier and you would have to install it on each project folder.
2. force to install globally, you could install using `sudo` before  npm install --global yarn, it will require the admin password of your mac, this is actually not recommended as npm packages would never be installed using sudo privileges.
3. uninstall node, and reinstall it using nvm which allows you to install npm packages globally without sudo privileges.

Best regards, Cromewar.

**Reply by Daruma1800** [reply]
@cromewar Do you have any guides for option 3? I am on windows and using WSL and getting the same error.

**Reply by cromewar** [reply]
In the course index "windows support" you could find a Guido about how you get it done with WSL.

**Reply by Daruma1800** [reply]
I found this video helpful https://www.youtube.com/watch?v=S_0pa93DVm8

**Comment by CrispyButter** [comment]
Thanks Cromewar. Which of the 3 solutions would you choose moving forward? I plan to create projects in the future to build a portfolio.

Also would you recommend I try to change the permissions? I am using my personal laptop so I am not sure why I wouldn't have permission to install npm packages.. Thanks again!

**Reply by cromewar** [reply]
Hi  @CrispyButter sorry for the delay (rough day), well I strongly suggest the option number 3 as you'll have some global packages you'll use for a lot of programs. I'm not a MAC user but [here](https://www.newline.co/@Adele/how-to-install-nodejs-and-npm-on-macos--22782681) you have  guide about how to install node using NVM or homebrew.

Why is not recommended to use `sudo`?

This could be a risk for you, may be installed trusted libraries like openzeppelin-contracts or usedapps is not a risk, but lets suppose you continue your career as developer and in some point you find and npm package you don't really know... if you install it using sudo privileges, the package will be able to execute and do almost anything on your computer, so short answer: using sudo for npm opens a door for hackers.

**Comment by snowieoxs** [comment]
found this video and seems to have got me going for now! ty

https://www.youtube.com/watch?v=M576WGiDBdQ

**Comment by cromewar** [comment, marked as answer]
Closing the thread due to owner's absence.
